File: crates/fey_math/src/vec3.rs
use crate::{Num, Signed, Vec2, Vec4, impl_vec, vec2, vec4};
use std::fmt::{Display, Formatter};

pub type Vec3F = Vec3<f32>;
pub type Vec3I = Vec3<i32>;
pub type Vec3U = Vec3<u32>;

/// A 3-dimensional vector.
#[repr(C)]
#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub struct Vec3<T> {
    pub x: T,
    pub y: T,
    pub z: T,
}

/// Create a [`Vec3`].
#[inline]
pub const fn vec3<T>(x: T, y: T, z: T) -> Vec3<T> {
    Vec3 { x, y, z }
}

impl_vec!(
    NAME = Vec3
    SHORT = vec3
    LEN = 3
    FIELDS = (x, y, z)
    TUPLE = (T, T, T)
);

impl<T> Vec3<T> {
    /// Return the vector with the x-value replaced.
    #[inline]
    pub fn with_x(self, x: T) -> Self {
        Self { x, ..self }
    }

    /// Return the vector with the y-value replaced.
    #[inline]
    pub fn with_y(self, y: T) -> Self {
        Self { y, ..self }
    }

    /// Return the vector with the y-value replaced.
    #[inline]
    pub fn with_z(self, z: T) -> Self {
        Self { z, ..self }
    }

    /// Add a fourth dimension to this vector.
    #[inline]
    pub fn with_w(self, w: T) -> Vec4<T> {
        vec4(self.x, self.y, self.z, w)
    }

    /// Swizzle components.
    #[inline]
    pub fn xzy(self) -> Self {
        vec3(self.x, self.z, self.y)
    }

    /// Swizzle components.
    #[inline]
    pub fn yxz(self) -> Self {
        vec3(self.y, self.x, self.z)
    }

    /// Swizzle components.
    #[inline]
    pub fn zxy(self) -> Self {
        vec3(self.z, self.x, self.y)
    }

    /// Swizzle components.
    #[inline]
    pub fn yzx(self) -> Self {
        vec3(self.y, self.z, self.x)
    }

    /// Swizzle components.
    #[inline]
    pub fn zyx(self) -> Self {
        vec3(self.z, self.y, self.x)
    }
}

impl<T: Copy> Vec3<T> {
    /// Reduce to a 2D vector by dropping `z`.
    #[inline]
    pub fn xy(&self) -> Vec2<T> {
        vec2(self.x, self.y)
    }
}

impl<T: Num> Vec3<T> {
    /// A unit vector representing the x-axis, equal to `(1, 0, 0)`.
    pub const X_AXIS: Self = vec3(T::ONE, T::ZERO, T::ZERO);

    /// A unit vector representing the y-axis, equal to `(0, 1, 0)`.
    pub const Y_AXIS: Self = vec3(T::ZERO, T::ONE, T::ZERO);

    /// A unit vector representing the z-axis, equal to `(0, 0, 1)`.
    pub const Z_AXIS: Self = vec3(T::ZERO, T::ZERO, T::ONE);

    /// A normalized vector pointing right, equal to `(1, 0, 0)`.
    pub const RIGHT: Self = vec3(T::ONE, T::ZERO, T::ZERO);

    /// A normalized vector pointing up, equal to `(0, 1, 0)`.
    pub const UP: Self = vec3(T::ZERO, T::ONE, T::ZERO);

    /// A normalized vector pointing forward, equal to `(0, 0, 1)`.
    pub const FORWARD: Self = vec3(T::ZERO, T::ZERO, T::ONE);
}

impl<T: Signed> Vec3<T> {
    /// A normalized vector pointing left, equal to `(-1, 0, 0)`.
    pub const LEFT: Self = vec3(T::NEG_ONE, T::ZERO, T::ZERO);

    /// A normalized vector pointing down, equal to `(0, -1, 0)`.
    pub const DOWN: Self = vec3(T::ZERO, T::NEG_ONE, T::ZERO);

    /// A normalized vector pointing backward, equal to `(0, 0, -1)`.
    pub const BACKWARD: Self = vec3(T::ZERO, T::ZERO, T::NEG_ONE);

    /// Returns the cross-product of this vector and another,
    /// equal to:
    ///
    /// `(y¹z²-y²z¹, -(x¹z²-x²z¹), x¹y²-x²y¹)`.
    #[inline]
    pub fn cross(self, other: Self) -> Self {
        vec3(
            self.y * other.z - other.y * self.z,
            -(self.x * other.z - other.x * self.z),
            self.x * other.y - other.x * self.y,
        )
    }

    /// Given the triangle `(a, b, c)`, and the interpolation values
    /// `ab` and `bc`, returns a barycentric coordinate.
    #[inline]
    pub fn barycentric(a: Self, b: Self, c: Self, t1: T, t2: T) -> Self {
        #[inline]
        fn coord<T: Signed>(a: T, b: T, c: T, ab: T, bc: T) -> T {
            a + (b - a) * ab + (c - a) * bc
        }
        vec3(
            coord(a.x, b.x, c.x, t1, t2),
            coord(a.y, b.y, c.y, t1, t2),
            coord(a.z, b.z, c.z, t1, t2),
        )
    }
}

impl<T: Display> Display for Vec3<T> {
    #[inline]
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        self.x.fmt(f)?;
        f.write_str(", ")?;
        self.y.fmt(f)?;
        f.write_str(", ")?;
        self.z.fmt(f)
    }
}

impl<T: Num> From<Vec2<T>> for Vec3<T> {
    #[inline]
    fn from(Vec2 { x, y }: Vec2<T>) -> Self {
        vec3(x, y, T::ZERO)
    }
}

File: crates/fey_math/lua/Vec3.lua
---@meta

---A 3-dimensional vector
---@class (exact) Vec3: Vec3Methods
---@field x number
---@field y number
---@field z number
---@operator add(Vec3): Vec3
---@operator sub(Vec3): Vec3
---@operator mul(Vec3|number): Vec3
---@operator div(Vec3|number): Vec3
---@operator unm: Vec3

---@class Vec3Class : Vec3Methods
---@overload fun(x: number, y: number, z: number): Vec3
local module = {}

---@class Vec3Methods
local methods = {}

---`(0, 0, 0)`
---@return Vec3
---@nodiscard
function module.zero() end

---`(1, 1, 1)`
---@return Vec3
---@nodiscard
function module.one() end

---`(1, 0, 0)`
---@return Vec3
---@nodiscard
function module.x_axis() end

---`(0, 1, 0)`
---@return Vec3
---@nodiscard
function module.y_axis() end

---`(0, 0, 1)`
---@return Vec3
---@nodiscard
function module.z_axis() end

---`(1, 0, 0)`
---@return Vec3
---@nodiscard
function module.right() end

---`(-1, 0, 0)`
---@return Vec3
---@nodiscard
function module.left() end

---`(0, -1, 0)`
---@return Vec3
---@nodiscard
function module.down() end

---`(0, 1, 0)`
---@return Vec3
---@nodiscard
function module.up() end

---`(0, 0, 1)`
---@return Vec3
---@nodiscard
function module.forward() end

---`(0, 0, -1)`
---@return Vec3
---@nodiscard
function module.backward() end

---Create a new vector.
---@param x number
---@param y number
---@param z number
---@return Vec3
---@nodiscard
function module.new(x, y, z) end

---Create a new vector `(val, val, val)`.
---@param val number
---@return Vec3
---@nodiscard
function module.splat(val) end

---Given the triangle `(a, b, c)`, and the interpolation values `ab` and `bc`, returns a barycentric coordinate.
---@param a Vec3
---@param b Vec3
---@param c Vec3
---@param ab number
---@param bc number
---@return Vec3
---@nodiscard
function module.barycentric(a, b, c, ab, bc) end

---Returns a temporary copy of this value.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.clone(self) end

---Boxes the value if not already boxed.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.box(self) end

---Boxes a copy of this value.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.box_clone(self) end

---Returns a copy of this vector with absolute value of all components.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.abs(self) end

---Returns `true` if the two vectors are approximately equal.
---@param self Vec3
---@param other Vec3
---@return boolean
---@nodiscard
function methods.approx(self, other) end

---Returns `true` if the vector is approximately equal to `(0, 0, 0)`.
---@param self Vec3
---@return boolean
---@nodiscard
function methods.approx_zero(self) end

---Rounds the vector's components up to the nearest integer.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.ceil(self) end

---Clamp's the vector's components between the components of `min` and `max`.
---@param self Vec3
---@param min Vec3
---@param max Vec3
---@return Vec3
---@nodiscard
function methods.clamp(self, min, max) end

---Returns the cross product of two vectors.
---@param self Vec3
---@param other Vec3
---@return Vec3
---@nodiscard
function methods.cross(self, other) end

---Returns the normalized direction towards the point.
---@param self Vec3
---@param point Vec3
---@return Vec3
---@nodiscard
function methods.dir_to(self, point) end

---Returns the distance between two points.
---@param self Vec3
---@param point Vec3
---@return Vec3
---@nodiscard
function methods.dist(self, point) end

---Returns the dot product of two vectors.
---@param self Vec3
---@param other Vec3
---@return number
---@nodiscard
function methods.dot(self, other) end

---Rounds the vector's components down to the nearest integer.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.floor(self) end

---Returns `true` if the vector is equal to `(0, 0, 0)`.
---@return boolean
---@nodiscard
function methods.is_zero(self) end

---Length of the vector.
---@param self Vec3
---@return number
---@nodiscard
function methods.len(self) end

---Returns a vector with components set to the largest values in the arguments.
---@param self Vec3
---@param ... Vec3
---@return Vec3
---@nodiscard
function methods.max(self, ...) end

---Returns a vector with components set to the smallest values in the arguments.
---@param self Vec3
---@param ... Vec3
---@return Vec3
---@nodiscard
function methods.min(self, ...) end

---Normalizes the vector.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.norm(self) end

---Reflects the vector off the provided normal.
---@param self Vec3
---@param normal Vec3
---@return Vec3
---@nodiscard
function methods.reflect(self, normal) end

---Rounds the vector's components to the nearest integer.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.round(self) end

---Returns a copy of this vector with all components signed (set to `1`, `-1`, or `0`).
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.sign(self) end

---Returns the square between two points.
---@param self Vec3
---@param point Vec3
---@return Vec3
---@nodiscard
function methods.sqr_dist(self, point) end

---Squared length of the vector.
---@param self Vec3
---@return number
---@nodiscard
function methods.sqr_len(self) end

---Returns a copy of this vector with all components rounded towards zero.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.trunc(self) end

---Changes the length of the vector without changing its direction.
---@param self Vec3
---@param new_len number
---@return number
---@nodiscard
function methods.with_len(self, new_len) end

---Replaces the `x` component of the vector.
---@param self Vec3
---@param x number
---@return Vec3
---@nodiscard
function methods.with_x(self, x) end

---Replaces the `y` component of the vector.
---@param self Vec3
---@param y number
---@return Vec3
---@nodiscard
function methods.with_y(self, y) end

---Replaces the `z` component of the vector.
---@param self Vec3
---@param z number
---@return Vec3
---@nodiscard
function methods.with_z(self, z) end

---Add a fourth dimension to this vector.
---@param self Vec3
---@param w number
---@return Vec3
---@nodiscard
function methods.with_w(self, w) end

---Returns the 2D portion of this vector.
---@param self Vec3
---@return Vec2
---@nodiscard
function methods.xy(self) end

---Swizzles the vector.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.xzy(self) end

---Swizzles the vector.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.yxz(self) end

---Swizzles the vector.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.yzx(self) end

---Swizzles the vector.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.zxy(self) end

---Swizzles the vector.
---@param self Vec3
---@return Vec3
---@nodiscard
function methods.zyx(self) end

return module
